### ISSUE TITLE
unit testing with vitest

### ISSUE THREAD

**Issue by zewa666** [ghiscoding/slickgrid-universal#1601]
### Describe the bug

hey there @ghiscoding

so I'm trying to get my unit tests up and running but noticed a few gotchas. one of them is that vitest loads slickgrid-common in combo with angular-slickgrid as cjs instead of esm. I'll try to understand where that comes from and fix it.

but while at it I noticed that the Sortablejs import now is wrong as the library tries to check the sortablejs.default instead of sortablejs variable. I guess this is due to non-named imports happening in slickgrid.ts

furthermore there is at least one more new place where jsdom causes issues, but I got a workaround for that which I'll add next week.

I'll also create a sample repo as this might involve a couple of more cases

### Reproduction

create unit tests for a grid using vite+vitest

### Which Framework are you using?

Angular

### Environment Info

```shell
| Executable          | Version |
| ------------------- | ------- |
| (framework used)    | VERSION |
| Slickgrid-Universal | VERSION |
| TypeScript          | VERSION |
| Browser(s)          | VERSION |
| System OS           | VERSION |
```

### Validations

- [X] Follow our [Code of Conduct](https://github.com/ghiscoding/slickgrid-universal/blob/master/.github/CODE_OF_CONDUCT.md)
- [X] Read the [Wikis](https://github.com/ghiscoding/slickgrid-universal/wiki).
- [X] Check that there isn't [already an issue](https://github.com/ghiscoding/slickgrid-universal/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/ghiscoding/slickgrid-universal/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by zewa666** [comment]
here is a related ticket for the cjs/esm loading https://github.com/vitest-dev/vitest/discussions/4233

**Comment by zewa666** [comment]
so I managed to force vitest to load up the esm instead of cjs version with the following config:

```ts
/// <reference types="vitest" />
import { defineConfig } from "vite";
import path from "path";

import angular from "@analogjs/vite-plugin-angular";

// HELPER TO BUILD PATHS TO ESM VERSION
function getAliasedPath(alias: string) {
  return path.resolve(
    __dirname,
    `../node_modules/@slickgrid-universal/${alias}/dist/esm/index.js`
  );
}

export default defineConfig(({ mode }) => ({
  plugins: [angular()],
  test: {
    globals: true,
    setupFiles: ["./src/test-setup.ts"],
    environment: "jsdom",
    include: ["src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}"],
    reporters: ["default", "junit"],
    coverage: {
      provider: "v8",
      reporter: ["cobertura", "text", "html"],
      reportsDirectory: "./reports"
    },
    outputFile: {
      junit: "./reports/junit.xml"
    },
    css: true,
    // REMOUNT CJS TO ESM REQUESTS
    alias: {
      "@slickgrid-universal/common": getAliasedPath("common"),
      "@slickgrid-universal/row-detail-view-plugin": getAliasedPath(
        "row-detail-view-plugin"
      ),
      "@slickgrid-universal/empty-warning-component": getAliasedPath(
        "empty-warning-component"
      ),
      "@slickgrid-universal/custom-footer-component": getAliasedPath(
        "custom-footer-component"
      ),
      "@slickgrid-universal/pagination-component": getAliasedPath(
        "pagination-component"
      ),
      "@slickgrid-universal/custom-tooltip-plugin": getAliasedPath(
        "custom-tooltip-plugin"
      ),
      "@slickgrid-universal/event-pub-sub": getAliasedPath("event-pub-sub"),
      "@slickgrid-universal/excel-export": getAliasedPath("excel-export"),
      "@slickgrid-universal/odata": getAliasedPath("odata")
    },
    server: {
      deps: {
        inline: ["angular-slickgrid"]
      }
    }
  },
  define: {
    "import.meta.vitest": mode !== "production"
  }
}));
```

The other issue that I mentioned was here https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/core/slickGrid.ts#L2796

where rule.left and rule.right are undefined due to the special treatment of cssRules which isn't compatible with JSDOM. So adding an if clause there fixes my issue:

```js
if (rule.left) {
  rule.left.style.left = `${x}px`;
}
if (rule.right) {
  rule.right.style.right = (((this._options.frozenColumn !== -1 && i > this._options.frozenColumn!) ? this.canvasWidthR : this.canvasWidthL) - x - w) + 'px';
}
```

**Comment by zewa666** [comment]
> > but while at it I noticed that the Sortablejs import now is wrong as the library tries to check the sortablejs.default instead of sortablejs variable. I guess this is due to non-named imports happening in slickgrid.ts
> 
> why would it do that? it's working fine in Jest and when using the App, so...? I also have `allowSyntheticDefaultImports` in the `tsconfig.base.json`, which is supposed to be enough

I think the issue might be due to a special treatment of those by Vitest itself:
https://vitest.dev/config/#deps-interopdefault

Anyways, I don't wont to care about those anyways since ESM is the way to go and the alias remaps as above work well enough for me without having to change anything on the slickgrid-universal side.

I guess documenting this in our test section should be good enough

**Comment by ghiscoding** [comment]
instead of the big change you've done in your Vitest config, you should still try the suggestion I proposed above. I have a good feeling that it will help since that `"node"` was exporting CJS first and you know orders matter in this case so that could explain the issue

**Comment by zewa666** [comment]
I tried messing around with the `exports` section and it didn't help. I guess the main reason is as stated in said conversation that the config of package.jsons is only inspected for top-level dependencies, so angular-slickgrid in this case, but not nested deps.

Anyways I've added a PR for the JSDOM fix as well as Doc updates for the Angular Repository

**Comment by ghiscoding** [comment]
another thing that I was never sure if we are supposed to put the `main` and `module` before or after the `exports`. I guess what would really help is some tools to find exactly the path that it takes to resolve. I think TSC has something for that, but Vitest most probably doesn't

**Comment by zewa666** [comment]
it would be really strange if the order of properties in a json obj matter as those anyways arent guaranteed. I that can be safely ignored

**Comment by ghiscoding** [comment]
even if you're surprised, I'm pretty sure that it's still a fact (at least for the `exports` where I know it totally matters), anyway I've opened PR #1603 with these changes, if it helps then great but if not, well I've tried 🙄

**Comment by zewa666** [comment]
js world never fails to surprise 🤪

up until recently you needed the full analogjs, but with a recent update now one can pick only the parts necessary for configuring vitest along with an exposed vite config.

**Comment by ghiscoding** [comment]
@zewa666 I've pushed a new release and that not only include your changes but also the changes I've done to cleanup the `exports` and other props in PR #1603 

One last and very important note, the latest version of Angular 18.1.x is giving problem caused by an indirect dependency `string-width` that was rewritten as ESM only a while ago but is now being imported by Angular 18.1.x causing CJS error thrown

> Unknown error: Error [ERR_REQUIRE_ESM]: require() of ES Module C:\OwnCloudDev10\GitHub\Angular-Slickgrid\node_modules\string-width\index.js from C:\OwnCloudDev10\GitHub\Angular-Slickgrid\node_modules\cliui\build\index.cjs not supported.
Instead change the require of index.js in C:\OwnCloudDev10\GitHub\Angular-Slickgrid\node_modules\cliui\build\index.cjs to a 
dynamic import() which is available in all CommonJS modules

I only have that problem in my Angular port, Aurelia & React are unaffected and again it's only after updating to 18.1.x. I also searched in the Angular GitHub project but couldn't find anything recent. I'm wondering if you're also having this problem? ... but anyway at the end of the day, my projects are not the cause of it and personally I don't like project that went for ESM only when an hybrid approach is a lot better and what I follow (support both CJS and prefer ESM)
